File: v6/prolog/labs/rel_spreading/probes/ts/c6_derived_source.ts
// C6 spreading a DERIVED source: what does a compiler need in order to splice
// columns out of something whose shape is itself computed?

declare function derive(): [id: number, tag: string];

// (1) TypeScript resolves the derived shape first, then splices. Accepted.
type DerivedRow = ReturnType<typeof derive>;
type SplicedFromDerived = [...DerivedRow, extra: number];
const ok: SplicedFromDerived = [1, "t", 7];

// (2) forward reference in file order: also accepted, the type pass is not
// source-ordered
type SplicedEarly = [...LaterRow, extra: number];
type LaterRow = [id: number, tag: string];
const okEarly: SplicedEarly = [1, "t", 7];

// (3) self reference through the splice: refused as circular
type SelfRow = [...SelfRow, extra: number];

// (4) mutual reference through the splice: refused as circular
type MutualA = [...MutualB, a: number];
type MutualB = [...MutualA, b: number];

export { ok, okEarly };
export type { SelfRow, MutualA, MutualB };

File: v6/prolog/labs/rel_spreading/probes/go/c6_derived_source.go
// C6 derived source in Go: can a declaration embed a shape that is itself
// computed, and what does a self-referential inclusion do?
package main

import "fmt"

type ARow struct {
	Id int64
}

// (1) embedding a named type declared LATER in the file: accepted, package
// scope is not source-ordered
type Early struct {
	Later
	Extra int64
}

type Later struct {
	Tag string
}

// (2) self-embedding: refused as an invalid recursive type
type SelfRow struct {
	SelfRow
	Extra int64
}

// (3) mutual embedding: refused as an invalid recursive type
type MutualA struct {
	MutualB
	A int64
}

type MutualB struct {
	MutualA
	B int64
}

func main() {
	e := Early{Later{"t"}, 7}
	fmt.Println(e.Tag, e.Extra)
	var s SelfRow
	var m MutualA
	fmt.Println(s, m, ARow{1})
}
